source/common/http/conn_manager_impl.cc

```diff
@@ -243,13 +243,11 @@ void ConnectionManagerImpl::doEndStream(ActiveStream& stream, bool check_for_def
   // here is when Envoy "ends" the stream by calling recreateStream at which point recreateStream
   // explicitly nulls out response_encoder to avoid the downstream being notified of the
   // Envoy-internal stream instance being ended.
-  if (stream.response_encoder_ != nullptr && (!stream.filter_manager_.remoteDecodeComplete() ||
+  if (stream.response_encoder_ != nullptr && (!stream.filter_manager_.decoderObservedEndStream() ||
                                               !stream.state_.codec_saw_local_complete_)) {
     // Indicate local is complete at this point so that if we reset during a continuation, we don't
     // raise further data or trailers.
     ENVOY_STREAM_LOG(debug, "doEndStream() resetting stream", stream);
-    // TODO(snowp): This call might not be necessary, try to clean up + remove setter function.
-    stream.filter_manager_.setLocalComplete();
```

Comment:
ping on this one

Reply:
I've looked through the code and the doEndStream is either called when local_complete_ was alraedy true (from filter manager) or the request is reset right down below which would result in callback to `ConnectionManagerImpl::ActiveStream::onResetStream()` which defer deletion of the ActiveStream with the filter manager in it. I suppose something can try calling FM after it was scheduled for deferred deletion, but I think it would be problem for other reasons.

I agree this is the most risky part of this PR. I can revert it and set the `decoder_filter_chain_aborted_` in this place, and then clean it up later with the runtime protection.

Reply:
yeah let's go with that.  I agree with the TODO we can probably get rid of this but I feel like it could use a runtime guard.  If you're willing to do it inline in this PR with a runtime guard that'd be even better because we can't forget to follow up, but OK either way

Reply:
Done